### PR TITLE
Convert options.js to TypeScript

### PR DESCRIPTION
**Pull request by supergrecko**
Migrates the options.js file to TypeScript, adding a generic `Record<string, unknown>` type for `window.compilerExplorerOptions`.

See also #2981

### CODE REVIEW

**Comment by partouf** [comment]
options.js is a special kind of file, I'm not sure if this actually works in production, so testing on staging first on this branch will be important

**Comment by supergrecko** [comment]
> options.js is a special kind of file, I'm not sure if this actually works in production, so testing on staging first on this branch will be important

Yeah, I'm a bit sceptical towards that \__webpack_public_path__ thing

File: static/options.ts [hunk 1]
```diff
@@ -22,18 +22,18 @@
...
-    window.compilerExplorerOptions[k] = extraOptions[k];
+const extraOptions: object = JSON.parse(decodeURIComponent(configElement.getAttribute('extraOptions')));
+for (const key in extraOptions) {
+    window.compilerExplorerOptions[key] = extraOptions[key];
 }
 
-__webpack_public_path__ = window.staticRoot;
+declare var __webpack_public_path__: string;
+
+__webpack_public_path__ = window.httpRoot;
```
Comment (apmorton) [inline static/options.ts:37]: Why did this change from `window.staticRoot` to `window.httpRoot`?
In production `staticRoot` will point to our CDN.

**Review by apmorton — CHANGES_REQUESTED** [review]

File: static/options.ts (hunk 1 above)
Comment (supergrecko) [reply static/options.ts]: Ups, that was not intended! Fixed.

Comment (supergrecko) [reply static/options.ts]: Ups, that was not intended! Fixed

**Comment by mattgodbolt** [comment]
@apmorton has this attended to your changes ok?

**Review by apmorton — APPROVED** [review]

**Review by mattgodbolt — APPROVED** [review]

**Pull request merged**